pkg/generator/fluentd/output/kafka/kafka.go

```diff
@@ -0,0 +1,175 @@
+package kafka
+
+import (
+	"net/url"
+	"strings"
+
+	logging "github.com/openshift/cluster-logging-operator/pkg/apis/logging/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	. "github.com/openshift/cluster-logging-operator/pkg/generator"
+	. "github.com/openshift/cluster-logging-operator/pkg/generator/fluentd/elements"
+	"github.com/openshift/cluster-logging-operator/pkg/generator/fluentd/helpers"
+	"github.com/openshift/cluster-logging-operator/pkg/generator/fluentd/output"
+	"github.com/openshift/cluster-logging-operator/pkg/generator/fluentd/output/security"
+	genhelper "github.com/openshift/cluster-logging-operator/pkg/generator/helpers"
+	urlhelper "github.com/openshift/cluster-logging-operator/pkg/generator/url"
+)
+
+const (
+	defaultKafkaTopic = "topic"
+)
+
+type Kafka struct {
+	Desc           string
+	StoreID        string
+	Brokers        string
+	Topics         string
+	SecurityConfig []Element
+	BufferConfig   []Element
+}
+
+func (k Kafka) Name() string {
+	return "kafkaTemplate"
+}
+
+func (k Kafka) Template() string {
+	return `{{define "` + k.Name() + `" -}}
+@type kafka2
+@id {{.StoreID}}
+brokers {{.Brokers}}
+default_topic {{.Topics}}
+use_event_time true
+{{- with $x := compose .SecurityConfig }}
+{{$x}}
+{{- end}}
+<format>
+  @type json
+</format>
+{{compose .BufferConfig}}
+{{- end}}
+`
+}
+
+func Conf(bufspec *logging.FluentdBufferSpec, secret *corev1.Secret, o logging.OutputSpec, op Options) []Element {
+	return []Element{
+		FromLabel{
+			InLabel: helpers.LabelName(o.Name),
+			SubElements: []Element{
+				Output(bufspec, secret, o, op),
+			},
+		},
+	}
+}
+
+func Output(bufspec *logging.FluentdBufferSpec, secret *corev1.Secret, o logging.OutputSpec, op Options) Element {
+	if genhelper.IsDebugOutput(op) {
+		return genhelper.DebugOutput
+	}
+	topics := Topics(o)
+	storeID := helpers.StoreID("", o.Name, "")
+	return Match{
+		MatchTags: "**",
+		MatchElement: Kafka{
+			StoreID:        strings.ToLower(helpers.Replacer.Replace(o.Name)),
+			Topics:         topics,
+			Brokers:        Brokers(o),
+			SecurityConfig: SecurityConfig(o, secret),
+			BufferConfig:   output.Buffer([]string{topics}, bufspec, storeID, &o),
+		},
+	}
+}
+
+//Brokers returns the list of broker endpoints of a kafka cluster.
+//The list represents only the initial set used by the collector's kafka client for the
+//first connention only. The collector's kafka client fetches constantly an updated list
+//from kafka. These updates are not reconciled back to the collector configuration.
+//The list of brokers are populated from the Kafka OutputSpec `Brokers` field, a list of
+//valid URLs. If none provided the target URL from the OutputSpec is used as fallback.
+//Finally, if neither approach works the current collector process will be terminated.
```

Comment:
This seems a little heavy-handed. This Kafka output can be one of many. Signaling this error in CLF status seems like a better option.

Reply:
this comment was lifted from old code as it is